src/main/java/com/alvarium/utils/SourceCodeUtilException.java

```diff
@@ -0,0 +1,21 @@
+
+/*******************************************************************************
+ * Copyright 2021 Dell Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+package com.alvarium.utils;
+
+public class SourceCodeUtilException extends Exception {
```

Comment:
Same with this class -- I don't think you need it. It's usage is mixed leading to semantic ambiguity. Sometimes it looks like this:

`throw new SourceCodeUtilException("Failed to read files", e);`

Other times it looks like this

`throw new SourceCodeUtilException("Hashing algorithm not found", e);`

So when the next developer comes along and tries to figure out what's going wrong based on the stack trace, the exception's type won't be of any help.

In the first case there's an exception in reading (or accessing a file). Might not using a vanilla File-related exception be better?

In the case of the `NotFoundException` a common pattern I see via searching is to subclass the `java.lang.RuntimeException`. Consider this level of specificity for other usages I may have missed.

Reply:
> Same with this class -- I don't think you need it. It's usage is mixed leading to semantic ambiguity. Sometimes it looks like this:
> 
> `throw new SourceCodeUtilException("Failed to read files", e);`
> 
> Other times it looks like this
> 
> `throw new SourceCodeUtilException("Hashing algorithm not found", e);`
> 
> So when the next developer comes along and tries to figure out what's going wrong based on the stack trace, the exception's type won't be of any help.
>
>In the first case there's an exception in reading (or accessing a file). Might not using a vanilla File-related exception be better?

What I had in mind is that the exception type would help the next developer understand that the exception was thrown during the checksum generation regardless of the exact component that fails which they can infer from the rest of the stack trace. Is this considered unnecessary ?

> In the case of the `NotFoundException` a common pattern I see via searching is to subclass the `java.lang.RuntimeException.` Consider this level of specificity for other usages I may have missed.

Are you referring to the `IOException` when the file isn't found ?

Reply:
> What I had in mind is that the exception type would help the next developer understand that the exception was thrown during the checksum generation regardless of the exact component that fails which they can infer from the rest of the stack trace. Is this considered unnecessary ?

Put that in the error message then. "Failed to read files. Could not create checksum." but make the exception type itself more specific.

> Are you referring to the IOException when the file isn't found ?
My specific comment w/r/t `NotFoundException` was for the "Hashing algorithm not found" case.

Reply:
> Put that in the error message then. "Failed to read files. Could not create checksum." but make the exception type itself more specific.

I wasn't sure if this was still an issue once we removed the `SourceCodeUtil` class. I followed the flow of other annotator implementations where all exceptions were wrapped with an `AnnotatorException` and I made an effort to make the messages more descriptive. If you consider this to still be problematic let me know.

> My specific comment w/r/t `NotFoundException` was for the "Hashing algorithm not found" case.
I've scanned the code for any `RuntimExceptions` and non seem to be a subclass of it. In the instance you mentioned, if the hashing algorithm isn't found on the system a `NoSuchAlgorithmException` is returned by the library which does not extend `RuntimeException` as well. Let me know if I misunderstood.